<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to perform eQTL analysis using MatrixEQTL</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xME"><tr><td>xME</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to perform eQTL analysis using MatrixEQTL</h2>

<h3>Description</h3>

<p><code>xME</code> is supposed to perform eQTL analysis using MatrixEQTL.
</p>


<h3>Usage</h3>

<pre>
xME(expression, genotype, GR.Gene, GR.SNP, mode = c("cis", "both",
"none"), distance.cis = 1e+06, pvalue.cis = 0.01,
pvalue.trans = 1e-05, num.PCs = 0, mydir = tempdir(check = T),
verbose = T, silent = F)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>expression</code></td>
<td>
<p>a data matrix/frame with genes in rows and samples in
columns</p>
</td></tr>
<tr valign="top"><td><code>genotype</code></td>
<td>
<p>a data (sparse) matrix/frame with SNPs in rows and
samples in columns</p>
</td></tr>
<tr valign="top"><td><code>GR.Gene</code></td>
<td>
<p>an GR object storing the genomic regions of genes</p>
</td></tr>
<tr valign="top"><td><code>GR.SNP</code></td>
<td>
<p>an GR object storing the genomic regions of SNPs</p>
</td></tr>
<tr valign="top"><td><code>mode</code></td>
<td>
<p>a character specifying the eQTL mode. It can be 'cis' for
cis-eQTL analysis, 'both' for both cis-eQTL and trans-eQTL analysis,
and 'none' (no eQTL analysis only returning variables actually used for
eQTL analysis: expression, genotype, GR.Gene and GR.SNP)</p>
</td></tr>
<tr valign="top"><td><code>distance.cis</code></td>
<td>
<p>a distance defining cis-eQTL. By default, it is
1Mb</p>
</td></tr>
<tr valign="top"><td><code>pvalue.cis</code></td>
<td>
<p>the p-value threshold outputing the cis-eQTL. By
default, it is 1e-2</p>
</td></tr>
<tr valign="top"><td><code>pvalue.trans</code></td>
<td>
<p>the p-value threshold outputing the trans-eQTL. By
default, it is 1e-5</p>
</td></tr>
<tr valign="top"><td><code>num.PCs</code></td>
<td>
<p>an integer specifying the number of principle components
(PCs) used for expression being regressed out. If zero (by default), no
gression is done</p>
</td></tr>
<tr valign="top"><td><code>mydir</code></td>
<td>
<p>a temporary directory file</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to TRUE for display</p>
</td></tr>
<tr valign="top"><td><code>silent</code></td>
<td>
<p>logical to indicate whether the messages will be silent
completely. By default, it sets to false. If true, verbose will be
forced to be false</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a &quot;MatrixEQTL&quot; object, a list including componets 'cis' and 'trans' (if
mode is 'both'). The 'cis' (or 'trans') contains a data frame 'eqtls'
with following columns
(&quot;snps&quot;,&quot;gene&quot;,&quot;statistic&quot;,&quot;pvalue&quot;,&quot;FDR&quot;,&quot;beta&quot;). If the mode is
'none', return a list with four components 'expression', 'genotype',
'GR.Gene' and 'GR.SNP'
</p>


<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xRegress</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(XGR)
res &lt;- xME(expression, genotype, GR.Gene, GR.SNP, mode='none')

## End(Not run)
</pre>


</body></html>
