<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to conduct colocalisation analysis through SMR</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xMEsmr"><tr><td>xMEsmr</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to conduct colocalisation analysis through SMR</h2>

<h3>Description</h3>

<p><code>xMEsmr</code> is supposed to conduct conduct Summary-data-based
Mendelian Randomisation (SMR) integrating GWAS and eQTL summary data.
</p>


<h3>Usage</h3>

<pre>
xMEsmr(gwas.summary, beqtl.summary, bfile, mode = c("both", "cis",
"trans"), peqtl = 5e-08, window.cis = 1000, window.trans = 1000,
heidi = T, heidi.peqtl = 0.00157, heidi.ld = c(0.05, 0.9),
heidi.num = c(3, 20), freq.check = F, thread.num = 1,
p.adjust.method = c("BH", "BY", "bonferroni", "holm", "hochberg",
"hommel"), verbose = T, silent = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gwas.summary</code></td>
<td>
<p>an input GWAS summary data file, containing columns
'snp', 'effect allele' (the allele assessed), 'other allele', 'freq'
(frequency of the effect allele; not essential unless 'freq.check' is
true), 'b' (effect size for the allele assessed; log(odds ratio) for a
case-control study), 'se' (standard error), 'p' (p-value) adn 'n'
(sample size; not required)</p>
</td></tr>
<tr valign="top"><td><code>beqtl.summary</code></td>
<td>
<p>a character specifying where to find eQTL summary
data files in the BESD format containing three files (.esi for SNP
information, .epi for probe information, and .besd for eQTL summary
statistics)</p>
</td></tr>
<tr valign="top"><td><code>bfile</code></td>
<td>
<p>a character specifying where to find the LD reference data
containing three files (.bed, .bim, and .fam)</p>
</td></tr>
<tr valign="top"><td><code>mode</code></td>
<td>
<p>a character specifying the SMR and HEIDI test mode. It can
be 'cis' for the test in cis regions, 'trans' for the test in trans
regions, and 'both' for both regions</p>
</td></tr>
<tr valign="top"><td><code>peqtl</code></td>
<td>
<p>eQTL p-value threshold for selecting a probe (with the top
associated eQTL passing a p-value threshold) for the SMR test. In other
words, a probe with the top associated eQTL not passing this threshold
will be removed for the test. By default, it is 5e-8</p>
</td></tr>
<tr valign="top"><td><code>window.cis</code></td>
<td>
<p>an integer specifying a window centred around the
probe to select cis-eQTLs (passing a p-value threshold) for the SMR
test. The default value is 1000Kb</p>
</td></tr>
<tr valign="top"><td><code>window.trans</code></td>
<td>
<p>an integer specifying a window centred around the
top associated trans-eQTL to select trans-eQTLs (passing a p-value
threshold) for the SMR and HEIDI test. The default value is 1000Kb</p>
</td></tr>
<tr valign="top"><td><code>heidi</code></td>
<td>
<p>logical to indicate whether the HEIDI test is enabled. By
default it is true</p>
</td></tr>
<tr valign="top"><td><code>heidi.peqtl</code></td>
<td>
<p>eQTL p-value threshold for selecting eQTLs per probe
for the HEIDI test. The default value is 1.57e-3 (a chi-squared value
10 with df=1)</p>
</td></tr>
<tr valign="top"><td><code>heidi.ld</code></td>
<td>
<p>LD r2 threshold used to further prune SNPs (eQTLs) in
the HEIDI test. By default only those between 0.05 and 0.9 will be used
for the test</p>
</td></tr>
<tr valign="top"><td><code>heidi.num</code></td>
<td>
<p>the number of SNPs (eQTLs) left per probe in the HEIDI
test. By default, the test skipped if the number of remaining SNPs is
less than 3; and top 20 SNPs (ranked by eQTL p-value) will be used in
the test</p>
</td></tr>
<tr valign="top"><td><code>freq.check</code></td>
<td>
<p>logical to indicate whether to remove SNPs withe
discrepant allele frequencies between data sets. By default it is
disabled</p>
</td></tr>
<tr valign="top"><td><code>thread.num</code></td>
<td>
<p>an integer specifying the number of OpenMP threads
for parallel computing. By default it is 1</p>
</td></tr>
<tr valign="top"><td><code>p.adjust.method</code></td>
<td>
<p>the method used to adjust p-values. It can be
one of &quot;BH&quot;, &quot;BY&quot;, &quot;bonferroni&quot;, &quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot;. The
first two methods &quot;BH&quot; (widely used) and &quot;BY&quot; control the false
discovery rate (FDR: the expected proportion of false discoveries
amongst the rejected hypotheses); the last four methods &quot;bonferroni&quot;,
&quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot; are designed to give strong control of
the family-wise error rate (FWER). Notes: FDR is a less stringent
condition than FWER</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to false for no display</p>
</td></tr>
<tr valign="top"><td><code>silent</code></td>
<td>
<p>logical to indicate whether the messages will be silent
completely. By default, it sets to false. If true, verbose will be
forced to be false</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a data frame with following columns (&quot;mode&quot;, &quot;probeID&quot;, &quot;Gene&quot;,
&quot;ProbeChr&quot;, &quot;Probe_bp&quot;, &quot;topSNP&quot;, &quot;topSNP_chr&quot;, &quot;topSNP_bp&quot;, &quot;A1&quot;,
&quot;A2&quot;, &quot;b_GWAS&quot;, &quot;b_eQTL&quot;, &quot;b_SMR&quot;, &quot;p_GWAS&quot;, &quot;p_eQTL&quot;, &quot;p_SMR&quot;,
&quot;fdr_SMR&quot;, &quot;p_HEIDI&quot;, &quot;fdr_HEIDI&quot;, &quot;nsnp_HEIDI&quot;)
</p>


<h3>Note</h3>

<p>None
</p>


<h3>See Also</h3>

<p><code>xMEsmr</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(XGR)

bfile &lt;- "~/Sites/SVN/github/bigdata_dev/Merged_EUR"
beqtl.summary &lt;-
"~/Sites/SVN/github/bigdata_dev/Pi_eQTL_hg19/Neutrophil"
gwas.summary &lt;- "summary_gwas.txt"
df_output &lt;- xMEsmr(gwas.summary, beqtl.summary, bfile)
utils::write.table(df_output, file="df_output.txt", row.names=F,
col.names=T, quote=F, sep="\t")

## End(Not run)
</pre>


</body></html>
